### PR TITLE
Hide green element storage mechanism in GreenNode::children

### PR DESCRIPTION
**Pull request by CAD97**
On top of #39. This is the main reason for splitting out new piecewise PRs, as this is the major breaking change required for changing the `GreenElement` internal packed representation from `enum NodeOrToken`.

r? @matklad

### CODE REVIEW

File: examples/s_expressions.rs [hunk 1]
```diff
@@ -313,8 +313,8 @@ impl Atom {
         Some(op)
     }
     fn text(&self) -> &SmolStr {
-        match &self.0.green().children()[0] {
-            rowan::NodeOrToken::Token(token) => token.text(),
+        match &self.0.green().children().nth(0) {
```
Comment (matklad) [inline examples/s_expressions.rs]: `.next()`

File: src/cursor.rs [hunk 2]
```diff
@@ -224,13 +223,12 @@ impl SyntaxNode {
...
                     .green()
                     .children()
-                    .iter()
                     .enumerate()
                     .map(|(i, child)| {
                         if i as u32 == me {
                             replacement.take().unwrap().into()
                         } else {
-                            child.clone()
+                            child.map(Clone::clone, Clone::clone)
```
Comment (matklad) [inline src/cursor.rs]: if that's the single use of `.map`, let's add `.cloned()` instead

File: src/green/node.rs [hunk 3]
```diff
@@ -1,7 +1,8 @@
 use super::*;
-use std::sync::Arc;
+use std::{slice, sync::Arc};
 
-use crate::{cursor::SyntaxKind, TextUnit};
+use crate::{cursor::SyntaxKind, NodeOrToken, TextUnit};
+use std::iter::FusedIterator;
```
Comment (matklad) [inline src/green/node.rs]: Nit: I try to stick with

```rust
use std::...;
// blank
use foo::...;
// blank
use crate::...;
// blank
use super::...;
```

**Comment by matklad** [comment]
bors r-

Hm, there's one thing I want to do here:let's bump version in Cargo.toml

**Comment by CAD97** [comment]
@matklad nits addressed

Let's remind myself whether I can do this...

bors: r+

**Comment by matklad** [comment]
grantid github perms. Accept the invite and ping me to kick bors perm sync

**Comment by CAD97** [comment]
@matklad accepted

**Comment by matklad** [comment]
Ook, try r+ again?

**Comment by CAD97** [comment]
bors: r=matklad

**Comment by CAD97** [comment]
@matklad before you publish, we should hide the use of `SmolStr` as well

**Comment by matklad** [comment]
@CAD97 I'd rather postpone that, as it would be a pretty big invasive change for rust-analyzer, and it would be interesting to measure the effects of tail-allocating the string *separatelly*.

**Comment by CAD97** [comment]
Got it!

**Comment by matklad** [comment]
https://github.com/rust-analyzer/rust-analyzer/pull/2314 updates to 0.7 in ra

**Pull request merged**